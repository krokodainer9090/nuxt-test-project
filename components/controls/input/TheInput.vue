<script>
const numberRegex = /^\d*$/;

export default {
  name: 'TheInput',

  props: {
    id: {
      type: Number,
      default: 1
    },
    value: {
      type: String,
      default: ''
    },
    type: {
      type: String,
      default: 'text'
    },
    min: {
      type: Number,
      default: 0
    }
  },

  computed: {
    isNumberInput () {
      return this.type === 'number';
    }
  },

  mounted () {
    if (this.isNumberInput) { this.numberEventListener(); }
  },

  destroyed () {
    if (this.isNumberInput) { this.numberEventListener(true); }
  },

  methods: {
    numberEventListener (unbind = false) {
      if (!this.$refs.input) { return; }

      if (unbind) {
        this.$refs.input.removeEventListener('keypress');

        return;
      }

      this.$refs.input.addEventListener('keypress', function (e) {
        const isNumber = numberRegex.test(e.target.value);

        if ( !e.target.value ) {
          e.target.value = 0;

          return;
        }

        if ( !isNumber ) {
          e.preventDefault();
        }
      });
    },

    getValue ({ target }) {
      this.$emit('input', target.value);
    }
  }
};
</script>

<template>
  <div class="input__container">
    <input
      :id="id"
      ref="input"
      :type="type"
      :value="value"
      :min="min"
      class="input"
      @input="getValue"
    >
  </div>
</template>

<style lang="scss">
@import "../../../styles/components/controls/the-input";
</style>
