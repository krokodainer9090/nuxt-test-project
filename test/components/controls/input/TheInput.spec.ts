// Components
import TheInput from '~/components/controls/input/TheInput.vue';

// Utilities
import { shallowMount, Wrapper } from '@vue/test-utils';

// Constants
const CLASS_NAME = '.input';
const newValue = '13';
const notNumberValue = 'gregre';

describe('TheInput', () => {
  const mountFunction: (options?: Object) => Wrapper<any> = (options = {}) => {
    return shallowMount(TheInput, {
      ...options
    });
  };

  it('should render input', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render number input with value', async () => {
    const wrapper = mountFunction({
      propsData: {
        type: 'number',
        value: '222'
      }
    });
    const emitObject = wrapper.emitted();

    // Check it's number input
    expect(wrapper.vm.isNumberInput).toBe(true);

    // Test input value emit
    const input = wrapper.find(CLASS_NAME);

    // @ts-ignore
    input.element.value = newValue;
    await input.trigger('input');

    expect(emitObject.input?.[0]).toEqual([newValue]);

    // Snapshot
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render number input and check keypress', async () => {
    const wrapper = mountFunction({
      propsData: {
        type: 'number',
        value: '222'
      }
    });
    const input = wrapper.find(CLASS_NAME);
    await input.trigger('keypress');

    // @ts-ignore
    input.element.value = notNumberValue;
    await input.trigger('keypress');
  });
});
