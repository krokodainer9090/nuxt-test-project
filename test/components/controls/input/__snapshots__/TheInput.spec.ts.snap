// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheInput should render input 1`] = `"<div class=\\"input__container\\"><input id=\\"1\\" type=\\"text\\" min=\\"0\\" class=\\"input\\"></div>"`;

exports[`TheInput should render number input with value 1`] = `"<div class=\\"input__container\\"><input id=\\"1\\" type=\\"number\\" min=\\"0\\" class=\\"input\\"></div>"`;
